### ISSUE TITLE
History incremental sync (between devices) / global suggestions

### ISSUE THREAD

**Issue by amerlyq** [qutebrowser/qutebrowser#6729]
## Usecases:

1) remember
* several devices and working environments:  home PC, home laptop, work PC, work laptop
* using device "A" you need open some link viewed several days ago on some another device "B/C/D" -- and you remember only keywords about that link title, nothing else

2) knowledge
* use timestamp when you opened some link/domain **for the first time** as unique identificator of that piece of information in personal knowledgebase. Of course you need queriable centralized DB for that.

3) archiving
* preserving single DB is always much easier than preserving DB per each device

**Comment by amerlyq** [comment]
Discussion we had long ago on IRC:
```
2020-05-26 22:41:01	amerlyq	1) How can I sync qutebrowser SQLite history between work/home/laptop PC ?
2020-05-26 22:41:39	amerlyq	2) How to sync history betweennnn different --basedir (for second "proxied" instance)?
2020-05-26 22:41:47	amerlyq	I haven't seen anything strictly similar in qutebrowser issues page.
2020-05-26 22:41:47	The-Compiler	amerlyq: 1) no idea - maybe something like Nextcloud works, though there might be conflicts
2020-05-26 22:42:12	The-Compiler	amerlyq: 2) Do you use both instances at the same time? Then you're pretty much out of luck
2020-05-26 22:44:13	amerlyq	I assume syncing would involve SQLite deserializing, merging and serializing back. Assuming all history entries have timestamps and are otherwise generally "unordered event lists", they must be easily mergable (i.e. grow-only history db).
2020-05-26 22:44:39	amerlyq	But desirializing my 500MiB history.db each time... is a maddness   
2020-05-26 22:45:59	amerlyq	So, no luck. Is there no similar proposal for "incremental sync", which you know of, The-Compiler?
2020-05-26 22:46:53	The-Compiler	amerlyq: not that I know of, and I'm sceptical it's worth the effort
2020-05-26 22:46:58	The-Compiler	(and complexity)
2020-05-26 22:49:00	amerlyq	Maybe. But I'm suffering from having no access to "auto-suggested" previous links I opened. Case #1: worked on some issues at work and must continue working on them at home through vpn. Case #2: started work on PC, continued working from laptop.
```

**Comment by amerlyq** [comment]
Piece of code I used for a period of time
```python
import sqlite3
from contextlib import closing

def qute_merge_others_into(dbpath, *others):
    cols = "url,title,atime,redirect"
    with closing(sqlite3.connect(dbpath)) as db, closing(db.cursor()) as cur:
        cur.execute(f"CREATE TABLE IF NOT EXISTS MergedHistory ({cols})")
        for path in (dbpath, *others):
            cur.execute(f"ATTACH '{path}' AS oth")
            cur.execute(
                f"""INSERT INTO MergedHistory ({cols})
                SELECT DISTINCT {cols} FROM oth.History"""
            )
            db.commit()
            cur.execute("DETACH oth")
        cur.execute("DELETE FROM CompletionHistory")
        cur.execute("UPDATE CompletionMetaInfo SET value=1 WHERE key=='force_rebuild'")
        cur.execute("DELETE FROM History")
        cur.execute(
            f"""INSERT INTO main.History ({cols})
            SELECT DISTINCT {cols} FROM MergedHistory"""
        )
        cur.execute("DROP TABLE MergedHistory")
        db.commit()
        cur.execute("VACUUM")
```


NOTE: better would be doing in-memory merging (i.e. open first db as `:memory:`, but then we must recreate/copy all tables and inject default values like user_version, hard to follow upstream changes :(
IDEA: extract tables/params SQL from .db and recreate new one based on them. Still subject to problems if upstream introduces foreigns or other high-level concepts/contexts over DB.

**Comment by amerlyq** [comment]
IDEA: incremental merge window -- based on last sync timestamp (per each synced device) 

NEED:
* each DB is guaranteed to be immutable append-only
*- old entries never changed and never deleted
*- can be loosened for shallow DB proxies, e.g. we keep single central "full" DB for archiving, and each device may removed some domains from its history, after push from DB was completed and verified
*- NEED: sep table with last_push.ts for stripping data below that ts
*- no new entries appear below last_sync.ts
* each DB must track/remember last sync date (BET: last entry) for each other DB
* ? how to prevent round-trips to sync prev synced state ?
```
      A <- fetch all new entries from B
       -> update last_sync.B=ts
      B <- fetch all unsynced entries from A
       -> update last_sync.B=ts
      A <- fetch B status, because B.last_entry= had become A.last_entry after prev sync
```

**Comment by amerlyq** [comment]
Basically, to implement local incremental sync (and maybe even firefox-like sync in future), we need some assurances:
* History table is immutable append-only
* user may create additional "last_pull" and "last_push" tables in DB (with list of user devices/hosts/instances of qute) and they won't be deleted by qutebrowser (or better Qutebrowser must not touch any other tables  beside which it states in some kind of "manifest")

Implementing future "online" continuous sync on top of that is as easy, as exchanging history increments (each 5min, onsize, ondemand, onclose) with some central daemon (simple python code), presumably running on user-controlled VPS.

@The-Compiler , how hard to ensure the two requirements above?
And overall feasibility of the plan?

**Comment by The-Compiler** [comment]
> @The-Compiler I strongly doubt that extension API will be enough for robust and usable syncing mechanics.

Care to elaborate?



> I can give up on "shallow history" but not on "originators" -- I will need this info locally when searching history in browser -- therefore, answering to @The-Compiler -- I doubt that extended api by themselves will cut it, without changes to qutebrowser.

Well, any approach will require changes to qutebrowser. The big question is how to change things in a way they are general enough so that the additional maintenance burden is worth it. Not every addition is worthwhile just because someone would love to have it :wink:

> Therefore waiting 2+ years for extapi would be nice

Interesting that you know timelines I don't.

> but nothing prevents us from having sync through direct db access in a week.

Yes, a lot of things do - for starters, a lot of assumptions about the on-disk format and qutebrowser's internal behavior which I'm not willing to commit to long-term. Hell, not even right now: Pressing Ctrl-D on a history item (`:completion-item-del`) deletes it from the `History` table.

**Comment by The-Compiler** [comment]
> Hm... on second glance looks solvable by extapi? But requires more api to be in "ready" state -- register command, register option, access history, plus many new api for WebHistory class for user metadata appendix.

Agreed.

> That's the whole point!
> We will use the direct access to DB -- no changes for Qute needed to maintain, beside comment "remember about sync", so we could look reprovingly if things break :D

Well then, there's nothing to do here? I'm confused. If that's the case, this probably should be turned into a discussion rather than an issue.

If you want (long-term) assurances about how qutebrowser is using the on-disk format, or changes of the on-disk format to make it suitable for direct access, my answer is likely going to be "no, sorry". But then your point about no additional maintenance burden is moot.

also pinging @lufte @rcorre @andrewcarlotti who did some history/SQL stuff and might have thoughts on this.
